sign/ed25519/ed25519.go

```diff
@@ -2,42 +2,93 @@ package ed25519
+// PublicKey represents a public key of Ed25519.
+type PublicKey [Size]byte
 
-// PrivKey represents a private key of Ed25519.
-type PrivKey [Size]byte
+// PrivateKey represents a private key of Ed25519. Implements crypto.Signer (golang.org/pkg/crypto/#Signer) interface.
+type PrivateKey struct {
+	public PublicKey  // This is a copy of the corresponding public key.
+	seed   [Size]byte // This is the secret (seed) key.
+}
 
 // Signature represents an Ed25519 signature.
 type Signature [2 * Size]byte
 
-// Pure corresponds to the Ed25519 PureEdDSA instance.
-type Pure struct{}
+// GetPublic returns a PublicKey corresponding to the private key.
+func (priv *PrivateKey) GetPublic() *PublicKey {
+	var public PublicKey
+	copy(public[:], priv.public[:])
+	return &public
+}
+
+// Public returns a crypto.PublicKey corresponding to the private key.
+func (priv *PrivateKey) Public() crypto.PublicKey { return priv.GetPublic() }
+
+// Sign signs the given message with priv.
+// Ed25519 performs two passes over messages to be signed and therefore cannot
+// handle pre-hashed messages. Thus opts.HashFunc() must return zero to
+// indicate the message hasn't been hashed. This can be achieved by passing
+// crypto.Hash(0) as the value for opts.
+func (priv *PrivateKey) Sign(rand io.Reader, message []byte, opts crypto.SignerOpts) ([]byte, error) {
+	if opts.HashFunc() != crypto.Hash(0) {
+		return nil, errors.New("ed25519: cannot sign hashed message")
+	}
+	sig := Sign(priv, message)
+	return sig[:], nil
+}
+
+// GenerateKey generates a pair of keys.
+func GenerateKey(rnd io.Reader) (*PrivateKey, *PublicKey, error) {
+	if rnd == nil {
+		rnd = rand.Reader
+	}
+
+	var seed [Size]byte
+	if _, err := io.ReadFull(rnd, seed[:]); err != nil {
+		return nil, nil, err
+	}
+	private := NewKeyFromSeed(seed[:])
+	return private, private.Public().(*PublicKey), nil
+}
 
-// KeyGen generates a public key from a secret key.
-func (e Pure) KeyGen(public *PubKey, private *PrivKey) {
-	k := sha512.Sum512(private[:])
+// NewKeyFromSeed calculates a private key from a seed. It will panic if
+// len(seed) is not SeedSize. This function is provided for interoperability
+// with RFC 8032. RFC 8032's private keys correspond to seeds in this
+// package.
+func NewKeyFromSeed(seed []byte) *PrivateKey {
+	if l := len(seed); l != Size {
+		panic("ed25519: bad seed length: " + strconv.Itoa(l))
+	}
+	var P pointR1
+	var private PrivateKey
+	k := sha512.Sum512(seed)
 	clamp(k[:])
 	reduceModOrder(k[:Size], false)
-	var P pointR1
 	P.fixedMult(k[:Size])
-	P.ToBytes(public[:])
+	P.ToBytes(private.public[:])
+	copy(private.seed[:], seed[:])
+	return &private
 }
 
 // Sign returns the signature of a message using both the private and public
 // keys of the signer.
-func (e Pure) Sign(message []byte, public *PubKey, private *PrivKey) *Signature {
-	k := sha512.Sum512(private[:])
+func Sign(private *PrivateKey, message []byte) Signature {
+	k := sha512.Sum512(private.seed[:])
```

Comment:
Since the original private key is not in use, would it make sense storing `s = H(k)` instead and avoid the recomputation here?

Reply:
`H(k)` is twice as large as `k` . Also hashing a short-length string it is not so painful.